### PR TITLE
Meghdad search

### PR DESCRIPTION
**Pull request by meghdadfm**
Rebase to develop

### CODE REVIEW

File: navigation/src/commonMain/kotlin/com/islamversity/navigation/Screens.kt [hunk 1]
```diff
@@ -27,8 +32,16 @@ sealed class Screens(
         EXTRA_SEARCH to SearchLocalModel.toByteArray(model),
         pushAnimation,
         popAnimation
-    ), Dynamic {
-        override val module: String
-            get() = "search"
-    }
+    )
+
+    class SoraDetail(
```
Comment (rezaiyan) [inline navigation/src/commonMain/kotlin/com/islamversity/navigation/Screens.kt]: Think of a proper name

File: search/src/main/res/layout/view_search.xml [hunk 2]
```diff
@@ -53,10 +53,14 @@
...
         app:layout_constraintLeft_toLeftOf="parent"
         app:layout_constraintRight_toRightOf="parent"
-        app:layout_constraintTop_toBottomOf="@+id/vSearchBack" />
+        app:layout_constraintTop_toBottomOf="@+id/vSearchBack"
+        tools:listitem="@layout/row_sura"
+        tools:itemCount="20"
+        />
 
 
 </androidx.constraintlayout.widget.ConstraintLayout>
```
Comment (rezaiyan) [inline search/src/main/res/layout/view_search.xml:66]: If it isn't necessary not to use ConstraintLayout in a row list.
It has performance issue!

**Comment by rezaiyan** [comment]
Well done.
Some of the classes are same as repo branch (quranHome merged)

**Review by rezaiyan — CHANGES_REQUESTED** [review]
Well done.
Some of the classes are same as develop branch (quranHome merged)

File: .idea/compiler.xml [hunk 3]
```diff
@@ -0,0 +1,27 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (rezaiyan) [inline .idea/compiler.xml]: The gitignore file has been updated in the develop branch

Comment (helloseyedjafari) [reply .idea/compiler.xml]: please rebase and remove this file from git

File: .idea/jarRepositories.xml [hunk 4]
```diff
@@ -0,0 +1,55 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (helloseyedjafari) [inline .idea/jarRepositories.xml]: same

File: core/src/commonMain/kotlin/com/islamversity/core/mvi/MviPresenter.kt [hunk 5]
```diff
@@ -37,6 +37,59 @@ abstract class BasePresenter<I : MviIntent, S : MviViewState, R : MviResult>(
     private val startState: S
 ) : MviPresenter<I, S> {
 
+
+    private val presenterJob = SupervisorJob()
```
Comment (helloseyedjafari) [inline core/src/commonMain/kotlin/com/islamversity/core/mvi/MviPresenter.kt:41]: please revert this file

File: daggercore/src/main/java/com/islamversity/daggercore/modules/domain/DomainModule.kt [hunk 6]
```diff
@@ -7,7 +7,9 @@ import dagger.Module
     includes = [
         SettingsModule::class,
         MappersModule::class,
-        LocalDataSourceModule::class
+        LocalDataSourceModule::class,
+        SearchRepoModule::class,
+        MappersModule::class
```
Comment (helloseyedjafari) [inline daggercore/src/main/java/com/islamversity/daggercore/modules/domain/DomainModule.kt]: MappersModule has been added twice

File: daggercore/src/main/java/com/islamversity/daggercore/modules/domain/SearchRepoModule.kt [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+
+@Module
+object SearchRepoModule {
+
+    @JvmStatic
+    @Provides
+    fun provideDataSource(
+        surahQueries : SurahQueries,
+        nameQueries : NameQueries
+    ): SurahLocalDataSource = SurahLocalDataSourceImpl(
```
Comment (helloseyedjafari) [inline daggercore/src/main/java/com/islamversity/daggercore/modules/domain/SearchRepoModule.kt]: This is already provided in the DBModule

File: db/src/commonMain/kotlin/com/islamversity/db/datasource/SoraLocalDataSource.kt [hunk 8]
```diff
@@ -0,0 +1,15 @@
+package com.islamversity.db.datasource
+
+import com.islamversity.db.model.Surah
+import kotlinx.coroutines.flow.Flow
+
+interface SoraLocalDataSource {
+    fun searchSora(query: String): Flow<List<Surah>>
+}
+class SoraLocalDataSourceImpl : SoraLocalDataSource {
```
Comment (helloseyedjafari) [inline db/src/commonMain/kotlin/com/islamversity/db/datasource/SoraLocalDataSource.kt]: This file is not necessary anymore

File: search/src/commonMain/kotlin/com/islamversity/search/SearchState.kt [hunk 9]
```diff
@@ -2,9 +2,11 @@ package com.islamversity.search
 
 import com.islamversity.core.mvi.BaseState
 import com.islamversity.core.mvi.BaseViewState
+import com.islamversity.domain.model.surah.SurahUIModel
 
 data class SearchState(
-    override val base: BaseState
+    override val base: BaseState,
+    val items : List<SurahUIModel>? = null
```
Comment (helloseyedjafari) [inline search/src/commonMain/kotlin/com/islamversity/search/SearchState.kt]: please use non-null list with `emptyList()` as default value

**Review by helloseyedjafari — COMMENTED** [review]
Thanks 
Almost done
please look at our comments

File: navigation/src/commonMain/kotlin/com/islamversity/navigation/Screens.kt (hunk 1 above)
Comment (meghdadfm) [reply navigation/src/commonMain/kotlin/com/islamversity/navigation/Screens.kt]: Seyed should suggest

File: search/src/main/res/layout/view_search.xml (hunk 2 above)
Comment (meghdadfm) [reply search/src/main/res/layout/view_search.xml:66]: This is just because we have not design yet

File: .idea/jarRepositories.xml (hunk 4 above)
Comment (meghdadfm) [reply .idea/jarRepositories.xml]: How ?

File: core/src/commonMain/kotlin/com/islamversity/core/mvi/MviPresenter.kt (hunk 5 above)
Comment (meghdadfm) [reply core/src/commonMain/kotlin/com/islamversity/core/mvi/MviPresenter.kt:41]: I don't understand

File: daggercore/src/main/java/com/islamversity/daggercore/modules/domain/SearchRepoModule.kt (hunk 7 above)
Comment (meghdadfm) [reply daggercore/src/main/java/com/islamversity/daggercore/modules/domain/SearchRepoModule.kt]: We didn't have that when I wrote it

File: db/src/commonMain/kotlin/com/islamversity/db/datasource/SoraLocalDataSource.kt (hunk 8 above)
Comment (meghdadfm) [reply db/src/commonMain/kotlin/com/islamversity/db/datasource/SoraLocalDataSource.kt]: Uhha

File: search/src/commonMain/kotlin/com/islamversity/search/SearchState.kt (hunk 9 above)
Comment (meghdadfm) [reply search/src/commonMain/kotlin/com/islamversity/search/SearchState.kt]: Ok, that is cause of rebase.

File: search/src/commonMain/kotlin/com/islamversity/search/model/SurahRowActionModel.kt [hunk 10]
```diff
@@ -0,0 +1,3 @@
+package com.islamversity.search.model
+
+class SurahRowActionModel(val selectedItem : SurahUIModel)
```
Comment (helloseyedjafari) [inline search/src/commonMain/kotlin/com/islamversity/search/model/SurahRowActionModel.kt]: Please use data class

File: search/src/main/res/layout/view_search.xml (hunk 2 above)
Comment (helloseyedjafari) [reply search/src/main/res/layout/view_search.xml:66]: Surah row and surah row xmls are the same right?

File: search/src/commonMain/kotlin/com/islamversity/search/SearchProcessor.kt [hunk 11]
```diff
@@ -1,12 +1,69 @@
...
-    override fun transformers(): List<FlowBlock<SearchIntent, SearchResult>> = listOf()
+    override fun transformers(): List<FlowBlock<SearchIntent, SearchResult>> = listOf(
+        newSearch,
+        itemClick
+    )
+
+    private val newSearch: FlowBlock<SearchIntent, SearchResult> = {
+        ofType<SearchIntent.Search>()
+            .flatMapLatest {
+                searchRepo.search(it.query, settingRepo.getCurrentSurahCalligraphy())
```
Comment (helloseyedjafari) [inline search/src/commonMain/kotlin/com/islamversity/search/SearchProcessor.kt]: Please consider using a usecase

File: search/src/androidMain/kotlin/com/islamversity/search/view/SurahRowView.kt [hunk 12]
```diff
@@ -0,0 +1,46 @@
...
+    attributeSet: AttributeSet? = null
+) : LinearLayout(context, attributeSet) {
+
+    private val binding = RowSuraBinding.inflate(LayoutInflater.from(context), this, true)
+
+    @ModelProp
+    lateinit var uiItem: SurahUIModel
+
+    @AfterPropsSet
+    fun bindView() {
```
Comment (helloseyedjafari) [inline search/src/androidMain/kotlin/com/islamversity/search/view/SurahRowView.kt]: We can merge this with modelProp

File: .gitignore [hunk 13]
```diff
@@ -33,8 +33,21 @@ captures/
 
 # IntelliJ
 *.iml
+<<<<<<< HEAD
```
Comment (helloseyedjafari) [inline .gitignore]: bad resolve

File: .idea/jarRepositories.xml (hunk 4 above)
Comment (helloseyedjafari) [reply .idea/jarRepositories.xml]: needs to be removed from git

File: domain/src/commonMain/kotlin/com/islamversity/domain/repo/surah/SurahSearchRepo.kt [hunk 14]
```diff
@@ -0,0 +1,30 @@
...
+    fun search(query: String, calligraphy: Calligraphy): Flow<List<SurahRepoModel>>
+}
+
+class SurahSearchRepoImpl(
+    private val dataSource: SurahLocalDataSource,
+    private val mapper: Mapper<Surah, SurahRepoModel>
+) : SurahSearchRepo {
+
+    override fun search(query: String, calligraphy: Calligraphy): Flow<List<SurahRepoModel>> =
+//        dataSource.observeAllSurahs(Calligraphy(LanguageCode("en") , CalligraphyName("should be ID") ))
```
Comment (helloseyedjafari) [inline domain/src/commonMain/kotlin/com/islamversity/domain/repo/surah/SurahSearchRepo.kt]: let's remove this comment now

File: search/src/androidMain/kotlin/com/islamversity/search/view/SearchView.kt [hunk 15]
```diff
@@ -92,23 +91,32 @@ class SearchView(
...
-                SearchIntent.NextPage(
-                    it.second.text.toString(),
-                    it.first.totalItemsCount,
-                    it.first.page
-                )
-            }
+
+    private fun itemClickIntent() = itemClickChannel
+        .receiveAsFlow()
+        .debounce(300.0.toDuration(DurationUnit.MILLISECONDS))
```
Comment (helloseyedjafari) [inline search/src/androidMain/kotlin/com/islamversity/search/view/SearchView.kt]: can we use throttleFIrst?

**Review by helloseyedjafari — COMMENTED** [review]
well done
final touches to go

**Review by helloseyedjafari — APPROVED** [review]

**Pull request merged**